### PR TITLE
Always install Rust and use `Leafwing-Studios/cargo-cache`

### PR DESCRIPTION
**Pull request by BD103**
This PR has two parts:

1. Some jobs don't actually install Rust before using it. This doesn't error because [Github's hosted runners come pre-installed with Rust](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2204-Readme.md#rust-tools), though it is frequently outdated.
2. Use [`Leafwing-Studios/cargo-cache`](https://github.com/Leafwing-Studios/cargo-cache) for all jobs that use Rust. This action is wonderful: I've been using it successfully for a month or so now. (Also it's by our friends at Leafwing, so it can be trusted!) This should speed up many jobs without the usual tradeoffs of caching.

### CODE REVIEW

**Comment by BD103** [comment]
I'm marking this as contentious because it is pulling in an outside dependency, but I feel like it's from a trusted source. If this gets major pushback, we can fork [`Leafwing-Studios/cargo-cache`](https://github.com/Leafwing-Studios/cargo-cache) under the Bevy organization and frequently update it.

**Comment by BD103** [comment]
I'm rebasing this on top of #1403 so CI passes. Consider this blocked until that PR is merged.

**Comment by rparrett** [comment]
Just linking in discussion from the twin PR for the `bevy` repo. https://github.com/bevyengine/bevy/pull/13040

**Comment by alice-i-cecile** [comment]
Yeah no objection to handing off ownership to the Bevy org if this appeals to y'all :)

**Review by TimJentzsch — APPROVED** [review]

**Review by janhohenheim — APPROVED** [review]

**Comment by BD103** [comment]
@mockersf what are your thoughts on merging this? It would provide a good testing ground for introducing this action to the main Bevy repository, and Alice even offered transferring the repository.

**Comment by janhohenheim** [comment]
Note that `cargo-cache` is about to release v2: https://github.com/Leafwing-Studios/cargo-cache/pull/17

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -36,6 +37,9 @@ jobs:
       - name: Install Rust
         uses: dtolnay/rust-toolchain@stable
 
+      - name: Cache Cargo files
+        uses: Leafwing-Studios/cargo-cache@v1
```
Comment (janhohenheim) [inline .github/workflows/ci.yml:41]:
```suggestion
        uses: Leafwing-Studios/cargo-cache@v2
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -49,6 +53,9 @@ jobs:
 
       - name: Install Rust
         uses: dtolnay/rust-toolchain@stable
+
+      - name: Cache Cargo files
+        uses: Leafwing-Studios/cargo-cache@v1
```
Comment (janhohenheim) [inline .github/workflows/ci.yml:58]:
```suggestion
        uses: Leafwing-Studios/cargo-cache@v2
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -92,6 +102,12 @@ jobs:
     steps:
       - uses: actions/checkout@v4
 
+      - name: Install Rust
+        uses: dtolnay/rust-toolchain@stable
+
+      - name: Cache Cargo files
+        uses: Leafwing-Studios/cargo-cache@v1
```
Comment (janhohenheim) [inline .github/workflows/ci.yml:109]:
```suggestion
        uses: Leafwing-Studios/cargo-cache@v2
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -127,15 +143,11 @@ jobs:
...
-            ~/.cargo/registry/index/
-            ~/.cargo/registry/cache/
-            ~/.cargo/git/db/
-            generate-errors/target/
-          key: ${{ runner.os }}-generate-errors-${{ hashFiles('generate-errors/Cargo.toml') }}
+      - name: Install Rust
+        uses: dtolnay/rust-toolchain@stable
+
+      - name: Cache Cargo files
+        uses: Leafwing-Studios/cargo-cache@v1
```
Comment (janhohenheim) [inline .github/workflows/ci.yml:150]:
```suggestion
        uses: Leafwing-Studios/cargo-cache@v2
```

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -158,6 +170,9 @@ jobs:
         with:
           target: wasm32-unknown-unknown
 
+      - name: Cache Cargo files
+        uses: Leafwing-Studios/cargo-cache@v1
```
Comment (janhohenheim) [inline .github/workflows/ci.yml:174]:
```suggestion
        uses: Leafwing-Studios/cargo-cache@v2
```

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -181,15 +196,11 @@ jobs:
...
-            ~/.cargo/registry/index/
-            ~/.cargo/registry/cache/
-            ~/.cargo/git/db/
-            generate-community/target/
-          key: ${{ runner.os }}-generate-community-${{ hashFiles('generate-community/Cargo.toml') }}
+      - name: Install Rust
+        uses: dtolnay/rust-toolchain@stable
+
+      - name: Cache Cargo files
+        uses: Leafwing-Studios/cargo-cache@v1
```
Comment (janhohenheim) [inline .github/workflows/ci.yml:203]:
```suggestion
        uses: Leafwing-Studios/cargo-cache@v2
```

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -61,6 +68,9 @@ jobs:
       - name: Install Rust
         uses: dtolnay/rust-toolchain@stable
 
+      - name: Cache Cargo files
+        uses: Leafwing-Studios/cargo-cache@v1
```
Comment (BD103) [inline .github/workflows/ci.yml:72]:
```suggestion
        uses: Leafwing-Studios/cargo-cache@v2
```

**Review by TrialDragon — APPROVED** [review]

**Comment by BD103** [comment]
Let me enable the `sweep-cache` option before this gets merged :)

**Comment by BD103** [comment]
I'm going to close this. I no longer have the time / energy to work on it, though anyone else is free to pick it up!

I'll create an issue for this, though, so it's not lost.

**Pull request closed without merging**